include/cuda/memory_resource

```diff
@@ -0,0 +1,309 @@
+#define _CUDA_MEMORY_RESOURCE
+
+#include <cuda/std/type_traits>
+
+#include <cuda/std/detail/__concepts>
+#include <cuda/std/detail/__config>
+
+#include <cuda/std/detail/__pragma_push>
+
+#if _LIBCUDACXX_STD_VER > 11
+_LIBCUDACXX_BEGIN_NAMESPACE_CUDA
+
+///////////////////////////////////////////////////////////////////////////////
+// memory_resource
+_LIBCUDACXX_TEMPLATE(class _Ret, class _Ptr)
+(requires _CUDA_VSTD::is_same_v<_Ptr, _Ret>) //
+    void _Returns_matching(_Ptr) {}
+
+/// \concept _Has_member_allocate
+/// \brief The \c _Has_member_allocate concept
+template <class _Resource>
+_LIBCUDACXX_CONCEPT_FRAGMENT(
+    _Has_member_allocate_,
+    requires(_Resource &__res, size_t __bytes, size_t __alignment) //
+    (_Returns_matching<void *>(__res.allocate(__bytes, __alignment))));
+template <class _Resource>
+_LIBCUDACXX_CONCEPT _Has_member_allocate =
+    _LIBCUDACXX_FRAGMENT(_Has_member_allocate_, _Resource);
+
+/// \concept _Has_member_deallocate
+/// \brief The \c _Has_member_deallocate concept
+template <class _Resource>
+_LIBCUDACXX_CONCEPT_FRAGMENT(_Has_member_deallocate_,
+                             requires(_Resource &__res, void *__ptr,
+                                      size_t __bytes,
+                                      size_t __alignment) //
+                             (__res.deallocate(__ptr, __bytes, __alignment)));
+template <class _Resource>
+_LIBCUDACXX_CONCEPT _Has_member_deallocate =
+    _LIBCUDACXX_FRAGMENT(_Has_member_deallocate_, _Resource);
+
+/// \concept resource
+/// \brief The \c resource concept
+template <class _Resource>
+_LIBCUDACXX_CONCEPT resource =
+    _Has_member_allocate<_Resource> &&_Has_member_deallocate<_Resource>
+    /* && _CUDA_VSTD::equality_comparable<_Resource> */;
+
+/// \concept has_property
+/// \brief The \c has_property concept
+template <class _Resource, class _Property>
+_LIBCUDACXX_CONCEPT_FRAGMENT(_Has_property_,
+                             requires(const _Resource &__res) //
+                             (get_property(__res, _Property{})));
+template <class _Resource, class _Property>
+_LIBCUDACXX_CONCEPT has_property = _LIBCUDACXX_FRAGMENT(_Has_property_,
+                                                        _Resource, _Property);
+
+/// \concept has_property_with
+/// \brief The \c has_property_with concept
+template <class _Resource, class _Property, class _Return>
+_LIBCUDACXX_CONCEPT_FRAGMENT(
+    _Has_property_with_,
+    requires(const _Resource &__res) //
+    (_Returns_matching<_Return>(get_property(__res, _Property{})),
+     _Returns_matching<_Return>(
+         _CUDA_VSTD::declval<typename _Property::value_type>())));
+template <class _Resource, class _Property, class _Return>
+_LIBCUDACXX_CONCEPT has_property_with =
+    _LIBCUDACXX_FRAGMENT(_Has_property_with_, _Resource, _Property, _Return);
+
+/// \concept resource
+/// \brief The \c resource concept
+template <class _Resource, class... _Properties>
+#if _LIBCUDACXX_STD_VER < 17
```

Comment:
There are quite some places, where we need to special case C++14 due to lack of fold expressions.

We should decide, whether there is a real benefit on C++14 support, or whether anyone that managed to leave C++11 behind should be expected to also have moved to C++17

Reply:
For just folding over `&&` ? We could have a macro for that.